### ISSUE TITLE
Simpler local_tempfile()

### ISSUE THREAD

**Issue by hadley** [r-lib/withr#141]
I had expected something like:

```R
local_tempfile <- function(envir = parent.frame()) {
  path <- tempfile()
  withr::defer(unlink(path), envir)
  
  path
}
```

Assignment into the calling environment seems a bit weird (I suspect it grew out of the `with_tempfile()` which doesn't have any other option)

**Comment by jimhester** [comment]
There are some usages of the function in the wild (https://github.com/search?l=R&p=2&q=%22local_tempfile%22&type=Code) and I don't really see a way to change to this style without breaking backwards compatibility.

It is also possible to create multiple temp files in one call with the current implementation, which would not be possible with the proposed approach.

The proposed implementation is a more intuitive API, I guess we could deprecate the existing one?

**Comment by hadley** [comment]
Yeah, with the new api you'd just call `local_tempfile()` multiple times if you wanted multiple temp files.

One possible way to change the API without breaking existing code would be to use the my proposal if the first argument was `NULL` (and deprecate that argument and remove in a few years)

**Comment by jimhester** [comment]
yeah, that is a good idea, we can do that

**Comment by jimhester** [comment]
fixed by 6a33f3b6e84d17

**Comment by gaborcsardi** [comment]
> One possible way to change the API without breaking existing code would be to use the my proposal if the first argument was NULL (and deprecate that argument and remove in a few years)

Cool, just came here to suggest this. :)
